### ISSUE TITLE
Customise Kernel Size?

### ISSUE THREAD

**Issue by Ryann2001** [SHI-Labs/NATTEN#146]
Thank you for your nice work. While reading the paper and using natten, I noticed that supported kernel size only included 3, 5, 7, 9, 11, 13, and can I customise a larger kernel size (such as 21 or larger) while using natten?

**Comment by AdityaKane2001** [comment]
@Ryann2001 

NATTEN already supports arbitrary kernel sizes (as long as they are odd). Here's an example:

```py
import torch
from natten.functional import na2d

q = torch.randn(4, 101, 101, 6, 64).cuda()
k = torch.randn(4, 101, 101, 6, 64).cuda()
v = torch.randn(4, 101, 101, 6, 64).cuda()


x = na2d(
    q,
    k,
    v,
    kernel_size=53,
    dilation=1,
)

print(x.shape) 
# Output: 
# torch.Size([4, 101, 101, 6, 64])
```

/cc @alihassanijr

**Comment by alihassanijr** [comment]
Thanks @AdityaKane2001 -- @Ryann2001 which version of NATTEN are you using?

The latest one comes with two choices:
1. Using the old unfused interface `na2d_qk`, and `na2d_av` which allows arbitrary kernel sizes (i.e. `3x7`, `7x5x9`, etc, as along as they're odd and > 1), dilations and causal masking. The only thing is that arbitrary kernel sizes or using causal masking goes down a slower path, so you might experience slower throughput in NA blocks. However, we have faster GEMM-based kernels, and for NA1D they allow arbitrary kernel sizes, and for NA2D they allow square kernels up to `63x63`, but there's no NA 3D implementations, those always fall back to the slow generic implementation.
2. Using the new [Fused Neighborhood Attention](https://arxiv.org/abs/2403.04690), which is what @AdityaKane2001 suggested above. Read more about its interface and how to use it here:
  * [Quick start guide](https://github.com/SHI-Labs/NATTEN/blob/main/docs/fna/fna-quickstart.md)
  * [Differences between fused and unfused in interface and API](https://github.com/SHI-Labs/NATTEN/blob/main/docs/fna/fused-vs-unfused.md)
  * [Auto-tuner](https://github.com/SHI-Labs/NATTEN/blob/main/docs/fna/autotuner.md)
  * [KV parallelism](https://github.com/SHI-Labs/NATTEN/blob/main/docs/fna/kv-parallelism.md)

Fused Neighborhood Attention is the fastest implementation, covers most recent NVIDIA architectures, is more memory efficient (similar implementation to Flash Attention), and offers NA1D, 2D and 3D, with arbitrary kernel sizes, dilation values, and causal mask layouts. We highly recommend using it instead.

**Comment by alihassanijr** [comment]
The issue appears to be resolved, so closing now. Feel free to reopen if you still have questions.
